doc/specs/#605 - Search/spec.md

```diff
@@ -0,0 +1,109 @@
+
+We will create a SearchBoxControl Xaml UserControl element. When a search process begins, a SearchBoxControl object will be created and attach to TermControl root grid. In other words, one SearchBox is added for each TermControl. The reasons for this design is:
+
+1. Each TermControl object is a Terminal Window and has a individual text buffer. In phrase 1 we are going to search witin the current terminal text buffer. 
+2. If we put the search box under TerminalApp, then the search can only happen on the current focused Terminal. 
+3. If the community does not like the current design, we can lift SearchBox to a higher level. 
+
+### Search process
+1. The user press Ctrl+F (or user's custom key binding) to open the search box. Focus will move to the TextBox. 
+2. Search is performed on a XAML TextBox. Once the user presses Enter or click up/down arrow button, we start to search from the last line of the current viewport or the current selection, and try to find the exact text in the text buffer. The nearest searched one will be selected. Then the search start point will be set to the selected text. The next search will start before or after the previous searched text.
+3. The user can choose to search up or down by choosing up arrow or down arrow buttons. The chosen button will be styled to indicate it is selected. If the user does not click the arrows buttons, the default direction is up. 
+4. The user can choose to do case sensitive or insensitive match by checking a check box. The default is case insensitive. 
+5. If the search box is focused, the user can navigte all the elements on the search box using tab. When selected, press Enter equals to click. 
+6. If the user click the "X" button or press <kbd>Esc</kbd>, the search stopped and the search box disappears and focus will move back to Terminal. In phrase one we do not store any state. 
+7. Once the search box is closed (exiting search mode), the selection will still be there. This coincides with the current VS Code and cmd experience. To get rid of the selection, the user can just click other area of the window.
+
+## UI/UX Design
+
+![SearchBox mockup](images/SearchBoxControl.png)
+
+Above is the SearchBoxControl. The two buttons with up/down arrows controls the search direction, it will be styled once clicked. The checkbox, if checked, means that we are searching case-sensitivily. The Dockbottom button on the right of the checkbox is for moving the search box to top/bottom. 
+
+![SearchBox mockup, arrow button clicked](images/SearchBoxUpSelected.png)

```

Comment:
From another lost comment:

> 
> the UI seems a little tall - one row for each of the close button, the search box, and the case sensitivity chckbox means that this UI is now obscuring 3 rows of terminal content. Perhaps we could collapse it to a single row more like VsCode?
>        ![image](https://user-images.githubusercontent.com/18356694/68127057-b553bb80-feda-11e9-99e5-6e64dea9c509.png)
> 
>  I would consider calling out these UI as _examples_ of what the UI could look like, but not necessarily what it will _definitely_ look like. Then we're less tied to content of the spec in whatever we end up implementing 😋

Reply:
I think if we put everything on one line, the search box would be very wide and it may block text more frequently ?

Reply:
Ooh, that's a good point. Though, the only real problem we'll have for it blocking text is for matches that are on the very first line of the buffer, right? Considering if the match is on any other line, the user can just scroll up to un-block text. 

Regardless, whatever the design is that's chosen, we should probably include arguments _for_ and _against_ each style in the spec, with a reason given for picking one style over the other.

Reply:
Added, and I think if the community provides feedbacks in the future that single-row layout is prefered, we can change  in the future.